pkg/prober/grpc.go

```diff
@@ -0,0 +1,54 @@
+package prober
+
+import (
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/thanos-io/thanos/pkg/component"
+	"google.golang.org/grpc/health"
+	grpc_health "google.golang.org/grpc/health/grpc_health_v1"
+)
+
+// GRPCProbe represents health and readiness status of given component, and provides GRPC integration.
+type GRPCProbe struct {
+	component component.Component
+	logger    log.Logger
+
+	h *health.Server
+}
+
+// NewGRPC creates a Probe that wrapped around grpc/healt.Server which reflects status of server.
+func NewGRPC(component component.Component, logger log.Logger) *GRPCProbe {
+	h := health.NewServer()
+	h.SetServingStatus("", grpc_health.HealthCheckResponse_NOT_SERVING)
+
+	return &GRPCProbe{component: component, logger: logger, h: h}
+}
+
+// HealthServer returns a gRPC health server which responds readiness and liveness checks.
+func (p *GRPCProbe) HealthServer() *health.Server {
+	return p.h
+}
+
+// Ready sets components status to ready.
+func (p *GRPCProbe) Ready() {
+	p.h.SetServingStatus("", grpc_health.HealthCheckResponse_SERVING)
+	level.Info(p.logger).Log("msg", "changing probe status", "status", "ready")
```

Comment:
Hm, wonder if it's consistent and clean to have log item added to gRPC methods but metrics added to HTTP methods. It's kind of weird. I would either:

* add single logging/metric prober (so prober responsible only to log/metric) and always combine all to the others
* add log line and metric to combine prober and use it everywhere. WDYT?

Reply:
I'll find a way to neatly combine observability layer for prober.